pkg/sql/compile/compile.go

```diff
@@ -357,32 +354,54 @@ func (c *Compile) compileApQuery(qry *plan.Query, ss []*Scope) (*Scope, error) {
 			Op:  vm.Deletion,
 			Arg: scp,
 		})
+		return rs, nil
 	case plan.Query_INSERT:
 		insertNode := qry.Nodes[qry.Steps[0]]
 		insertNode.NotCacheable = true
 
+		rs := c.newMergeScope(ss)
+		if len(insertNode.InsertCtx.OnDuplicateIdx) > 0 {
+			onDuplicateKeyArg, err := constructOnduplicateKey(insertNode, c.e, c.proc)
+			if err != nil {
+				return nil, err
+			}
+			rs.Instructions = append(rs.Instructions, vm.Instruction{
+				Op:  vm.OnDuplicateKey,
+				Arg: onDuplicateKeyArg,
+			})
+		}
 		preArg, err := constructPreInsert(insertNode, c.e, c.proc)
 		if err != nil {
 			return nil, err
 		}
+		rs.Instructions = append(rs.Instructions, vm.Instruction{
+			Op:  vm.PreInsert,
+			Arg: preArg,
+		})
+
+		if pkIdx, ok := getPKIdx(insertNode.InsertCtx.TableDef); ok {
+			dedupScope, err := buildDedupScope(c, pkIdx, insertNode.InsertCtx.TableDef.Cols[pkIdx])
+			if err != nil {
+				return nil, err
+			}
+			rs.PreScopes = append(rs.PreScopes, dedupScope)
+			rs.Instructions = append(rs.Instructions, vm.Instruction{
+				Op:  vm.Dispatch,
+				Idx: c.anal.curr,
+				Arg: constructDispatchLocal(true, dedupScope.Proc.Reg.MergeReceivers),
+			})
+		}
 
 		arg, err := constructInsert(insertNode, c.e, c.proc)
 		if err != nil {
 			return nil, err
 		}
-		nodeStats := qry.Nodes[insertNode.Children[0]].Stats
 
+		nodeStats := qry.Nodes[insertNode.Children[0]].Stats
 		if nodeStats.GetCost()*float64(SingleLineSizeEstimate) > float64(DistributedThreshold) || qry.LoadTag {
 			// use distributed-insert
 			arg.IsRemote = true
-			for _, scope := range ss {
-				scope.Instructions = append(scope.Instructions, vm.Instruction{
-					Op:  vm.PreInsert,
-					Arg: preArg,
-				})
-			}
-
-			rs = c.newInsertMergeScope(arg, preArg, ss)
+			rs = c.newInsertMergeScope(arg, ss)
```

Comment:
same with last comment. I think Insert should just appear after the group operator. And you also need a mergegroup operator to do duplicate-check